File: Language.Implementation.Patterns/cxx/ch02_basic_parsing_patterns/p02_lexer/test/LexerTest.cpp
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "Token.hpp"
#include "ListLexer.hpp"

TEST(Lexer, nextToken)
{
    std::string input = "[a, b ]";

    ListLexer lexer{input};
    std::vector<Token> expectedResult;
    expectedResult.push_back(Token{ListLexer::LBRACK, "["});
    expectedResult.push_back(Token{ListLexer::NAME, "a"});
    expectedResult.push_back(Token{ListLexer::COMMA, ","});
    expectedResult.push_back(Token{ListLexer::NAME, "b"});
    expectedResult.push_back(Token{ListLexer::RBRACK, "]"});

    std::vector<Token> actualResult;

    Token t = lexer.nextToken();
    while (t.type != Lexer::EOF_TYPE) {
        actualResult.push_back(t);
        t = lexer.nextToken();
    }

    ASSERT_EQ(expectedResult.size(), actualResult.size());

    for (int i = 0; i < expectedResult.size(); ++i) {
        ASSERT_EQ(expectedResult.at(i).text, actualResult.at(i).text);
        ASSERT_EQ(expectedResult.at(i).type, actualResult.at(i).type);
    }
}

File: Language.Implementation.Patterns/python/ch02_basic_parsing_patterns/p02_lexer_2/LexerTest.py
from Token import Token
from Lexer import Lexer
from ListLexer import ListLexer

def test_lexer():
    input = "[a, b ]"
    lexer = ListLexer(input)

    expectedResult = []
    expectedResult.append(Token(Token.LBRACK, "["))
    expectedResult.append(Token(Token.NAME, "a"))
    expectedResult.append(Token(Token.COMMA, ","))
    expectedResult.append(Token(Token.NAME, "b"))
    expectedResult.append(Token(Token.RBRACK, "]"))

    actualResult = []

    t = lexer.nextToken()
    while t.type != Token.EOF_TYPE:
        actualResult.append(t)
        t = lexer.nextToken()

    assert len(expectedResult) == len(actualResult)

    for i in range(len(expectedResult)):
        assert expectedResult[i].text == actualResult[i].text
        assert expectedResult[i].type == actualResult[i].type


if __name__ == "__main__":
    test_lexer()
